### PR TITLE
Avoid potential OOB if width > sizeof(start)

### PR DESCRIPTION
**Pull request by BenBE**
This can't currently happen due to `sizeof(start)` being way larger than `MAX_OPT_HELP_WIDTH`, but wasn't (properly) checked for previously. With this patch there still remains one (static) OOB, when the length of the option name and the `valtype2param` string for that argument together  overflow the buffer in `opt_print`. This is kinda unlikely, unless someone intentionally crafts a
long option name, in which case this would become some trivial stack buffer overrun with
possibility to overwrite the pointer to the `OPTIONS` structure (a long `o->name` is the critical path here). All current documentation for options seem to be sane here ATM.

I sincerely hope we trust our built-in documentation to not exploit ourselves.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by BenBE** [comment]
NB: Patch created based on the issue I commented on in #11918.

**Comment by slontis** [comment]
I think this is really another case of a false positive..
opt_print() is 'currently' only accessed by opt_help().
And the options are all a fixed size.

You will see that opt_help() also has a buffer  char start[80 + 1];
and does OPENSSL_assert(i < (int)sizeof(start));

So would it be sufficient to just change that code to do the following?

```
#define MAX_HELP_OPT_SIZE (80 + 1)

opt_print(...)
{
    char start[MAX_HELP_OPT_SIZE];

    OPENSSL_assert(width < (int)sizeof(start));
   ...
   ...
}

opt_help(...)
{
    char start[MAX_HELP_OPT_SIZE];
}

```

**Comment by BenBE** [comment]
No, as this won't fix the issue in case `MAX_OPT_HELP_WIDTH > MAX_OPT_HELP_SIZE`.

I see there is little incentive in fixing it as all inputs to those functions are static. But better do something properly and avoid problems in the future. I put the "I hope we trust our documentation" for a reason. ;-)

And as shown in the patches, there are two OOB issues here. The `width>sizeof(buf)` one seems unexploitable, as the necessary checks are there, but the second one (using the large `opt->name` is there and easily translates into RCE, if you just fancied to write some option string long enough to overwrite the function return address.

Also does the current setup allocate the memory buffer `start` twice on the stack (once in `opt_help` and once in `opt_print`). You can avoid inconsistencies between those routines by using a define, or avoid this source of inconsistency altogether by eliminating the second stack buffer, which is unused (apart from the length check).

My rewrite of those routines avoids those issues completely, and if format strings in `BIO_printf` support width formatting (`%*s`) you could even get rid of that `start` buffer completely AFAIC. Haven't checked this, but the patch for this would be easy.

IMO leaving the current implementation as is leaves a taste of bad code hygiene.

**Comment by mattcaswell** [comment]
> No, as this won't fix the issue in case MAX_OPT_HELP_WIDTH > MAX_OPT_HELP_SIZE.

Perhaps define `MAX_OPT_HELP_SIZE` in terms of `MAX_OPT_HELP_WIDTH` so that this can never be the case?

**Comment by BenBE** [comment]
> > No, as this won't fix the issue in case MAX_OPT_HELP_WIDTH > MAX_OPT_HELP_SIZE.
> 
> Perhaps define `MAX_OPT_HELP_SIZE` in terms of `MAX_OPT_HELP_WIDTH` so that this can never be the case?

Doesn't make much sense in my eyes, as they are independent of each other:
```
 -name_of_very_long_option_that_may_overflow_the_buffer_as_this_name_here_is_not_bounded_by_anything_in_the_option_definition_and_thus_can_be_extend_to_infinity_or_until_I_run_out_of_ideas_to_make_this_ridiculous_option_name_even_longer PEM|DER@ENGINE // <-- MAX_OPT_HELP_SIZE
<-- - - - - MAX_OPT_HELP_WIDTH - - - - -->  This is the help text for this fancy option
```

The first line overwrites (unchecked) in both strcat calls for the name AND the valtype.

Fixing the relation as suggested fixes the OOB noted in the first patch, while leaving the OOB that is fixed in the second patch.

FWIW: @richsalz suggested limitting length of `option->name` in `list_options_for_command` (https://github.com/openssl/openssl/blob/master/apps/list.c#L420) by causing a CI failure.

IMO this is a somewhat lazy approach, instead of fixing the underlaying issue(s).

**Comment by slontis** [comment]
Here is an alternative that uses the same code to calculate the width and do the print.
I think it is a bit cleaner?

```
static int put_char(char *out, char ch, int max_len, int *cur_index)
{
    if (*cur_index >= max_len)
        return 0;
    if (out != NULL)
        out[*cur_index] = ch;
    *cur_index += 1;
    return 1;
}

static int put_string(char *out, const char *str, int max_len, int *cur_index)
{
    int len = (int)strlen(str);
    int remain = max_len - 1 - *cur_index;

    if (remain <= 0)
        return 0;

    if (out != NULL) {
        if (remain < len)
            len = remain;
        memcpy(out + *cur_index, str, len);
    }
    *cur_index += len;
    return 1;
}

static int do_print(const OPTIONS *o, int doingparams, char *out, size_t out_len)
{
    int ret = 1, width = 0;
    const char *help;

    help = o->helpstr ? o->helpstr : "(No additional info)";
    if (o->name == OPT_HELP_STR) {
        if (out != NULL)
            opt_printf_stderr(help, prog);
        return 0;
    }
    if (o->name == OPT_SECTION_STR) {
        if (out != NULL) {
            opt_printf_stderr("\n");
            opt_printf_stderr(help, prog);
        }
        return 0;
    }
    if (o->name == OPT_PARAM_STR) {
        if (out != NULL)
            opt_printf_stderr("\nParameters:\n");
        return 0;
    }

    /* Fill with spaces */
    if (out != NULL) {
        memset(out, ' ', out_len - 1);
        out[out_len - 1] = '\0';
    }
    if (o->name == OPT_MORE_STR) {
        if (out != NULL)
            opt_printf_stderr("%s  %s\n", out, help);
        return 0;
    }

    ret = put_char(out, ' ', out_len, &width);
    if (!doingparams)
        ret = ret && put_char(out, '-', out_len, &width);

    if (o->name[0])
        ret = put_string(out, o->name, out_len, &width);
    else
        ret = put_char(out, '*', out_len, &width);

    if (o->valtype != '-') {
        ret = ret && put_char(out, ' ', out_len, &width);
        ret = ret && put_string(out, valtype2param(o), out_len, &width);
    }

    if (out != NULL)
        opt_printf_stderr("%s  %s\n", out, help);

    if (width != 0)
        width++;    /* Include the NUL terminator in the count */
    return width;
}

void opt_help(const OPTIONS *list)
{
    const OPTIONS *o;
    int sawparams = 0, w, width = 5;
    char buffer[MAX_OPT_HELP_WIDTH + 1];

    /* Starts with its own help message? */
    if (list[0].name != OPT_HELP_STR) {
        opt_printf_stderr("Usage: %s [options]\n", prog);
        if (list[0].name != OPT_SECTION_STR)
            opt_printf_stderr("Valid options are:\n", prog);
    }

    /* Find the widest help. */
    for (o = list; o->name; o++) {
        if (o->name == OPT_PARAM_STR)
            sawparams = 1;
        w = do_print(o, sawparams, NULL, MAX_OPT_HELP_WIDTH + 1);
        if (w > width)
            width = w;
    }

    /* Now let's print. */
    sawparams = 0;
    for (o = list; o->name; o++) {
        if (o->name == OPT_PARAM_STR)
            sawparams = 1;
        do_print(o, sawparams, buffer, width);
    }
}

```

**Comment by BenBE** [comment]
Well, in https://github.com/openssl/openssl/pull/12265/commits/0eb18590ac37d68a509ac0e441774f05128479b6 a changed the overall code in a way that completely gets rid of the string concatenation stuff.

Unfortunately it seems to have been stuck in the CI; if somebody could kick the CI to retest …

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
IMO we should just take this.

File: apps/lib/opt.c [hunk 1]
```diff
@@ -870,63 +870,67 @@ void opt_print(const OPTIONS *o, int doingparams, int width)
...
-            return;
-        }
-        if (o->name == OPT_PARAM_STR) {
-            opt_printf_stderr("\nParameters:\n");
-            return;
-        }
+    int linelen, printlen;
+
+    /* Avoid OOB if width is beyond the buffer size of start */
+    if(width >= (int)sizeof(start))
```
Comment (t8m) [inline apps/lib/opt.c:876]: Nit: - please add a space after `if`

**Review by t8m — CHANGES_REQUESTED** [review]
One minor nit

File: apps/lib/opt.c (hunk 1 above)
Comment (BenBE) [reply apps/lib/opt.c:876]: Fixed. ;-)

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for this.

**Pull request closed without merging**